modules/data.land/R/soil_params_ensemble.R

```diff
@@ -0,0 +1,224 @@
+##' silt <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/silt_percent.rds") #silt fraction in percentage
+##' settings <-read.settings("/projectnb/dietzelab/Cherry/xml/pecan_monthly_SDA_soilwater.xml")
+##' soil_params_ensemble(settings,sand,clay,silt,outdir)
+##'  }
+##'  
+##'  @return Ensemble soil parameter files defined in outdir and file paths in xml file
+##'  @export 
+##'  @author Qianyu Li
+##'  @importFrom magrittr %>%
+##'  
+
+soil_params_ensemble <- function(settings,sand,clay,silt,outdir,write_into_settings=TRUE){
+  
+  # A function to rescale the sums of mean texture fractions to 1 as the original sums are slightly different from 1 for some layers
+  rescale_sum_to_one <- function(sand, clay, silt) {
+    total <- sand + clay + silt
+    rescaled_sand <- sand / total
+    rescaled_clay <- clay / total
+    rescaled_silt <- silt / total
+    return(list(
+      sand = rescaled_sand,
+      clay = rescaled_clay,
+      silt = rescaled_silt))
+  }
+  
+  # A function to reformat the nested list as inputs to "soil2netcdf" function
+  reformat_soil_list <- function(samples_all_depth) {
+    # Define the fractions we want to extract
+    fractions <-
+      c("fraction_of_sand_in_soil",
+        "fraction_of_clay_in_soil",
+        "fraction_of_silt_in_soil")
+    
+    # Initialize a new list to store reformatted data
+    reformatted <-setNames(vector("list", length(fractions)), fractions)
+    
+    # Extract data for each fraction
+    for (fraction in fractions) {
+      reformatted[[fraction]] <-
+        unlist(lapply(samples_all_depth, function(depth_list) {
+          depth_list[[fraction]] # Extract the fraction value
+        })) %>% purrr::set_names(NULL)
+    }
+    # Combine depth into a single vector for readability
+    reformatted$soil_depth <- as.numeric(names(samples_all_depth))
+    return(reformatted)
+  }
+  
+  # A function to write to settings
+  create_mult_list <- function(list.names, paths) {
+    out <- as.list(paths)
+    names(out) <- list.names
+    out
+  }
+
+  # Convert values to proportion (0-1) from percentage
+  if (any(c(sand$Value, clay$Value, silt$Value) > 2)) {
+     sand$Value <- if (is.null(sand$Value)) { NULL } else { sand$Value / 100 }
+     clay$Value <- if (is.null(clay$Value)) { NULL } else { clay$Value / 100 }
+     silt$Value <- if (is.null(silt$Value)) { NULL } else { silt$Value / 100 }
+  }
+  ens_n <- settings$ensemble$size
+  # Merge all soil texture data together
+  texture_all <-merge(sand, clay, by=c("Depth", "Quantile", "Siteid"))  %>% merge(silt, by=c("Depth", "Quantile", "Siteid")) %>%
+  `colnames<-`(c(
+      "soil_depth", #"soil_depth" will be used in "soil2netcdf" function
+      "quantile",
+      "siteid",
+      "fraction_of_sand_in_soil",
+      "fraction_of_clay_in_soil",
+      "fraction_of_silt_in_soil"))
+
+   # Substitute the depth range with the middle depth value in meter
```

Comment:
Why middle? I'm pretty sure earlier soil code was working with the bottoms of each layer (with the assumption that the first layer's top is always at 0 by definition). Using the middles will generate ambiguity about layer thicknesses.

Reply:
Fixed with layer bottom value.